src/commonMain/kotlin/io/ktor/io/ByteArrayBuffer.kt

```diff
@@ -1,184 +1,115 @@
 package io.ktor.io
 
-import io.ktor.io.internal.*
 import kotlin.math.min
 
 public const val DEFAULT_POOL_CAPACITY: Int = 2000
 public const val DEFAULT_BUFFER_SIZE: Int = 1024 * 16
 
 public class ByteArrayBuffer(
-    public override val capacity: Int,
-    private val pool: ObjectPool<ByteArrayBuffer>
+    array: ByteArray,
+    private val pool: ObjectPool<ByteArray>
 ) : Buffer() {
 
-    public constructor(capacity: Int) : this(capacity, ByteArrayBufferPool.NoPool)
+    /**
+     * Creates buffer of fixed [capacity].
+     */
+    public constructor(capacity: Int) : this(ByteArray(capacity), ByteArrayPool.NoPool)
 
-    internal val array = ByteArray(capacity)
+    public constructor(pool: ObjectPool<ByteArray>) : this(pool.borrow(), pool)
+
+    /**
+     * Provides access to underlying byte array.
+     *
+     * Please note, all changes of the array will be reflected in the buffer.
+     */
+    @Suppress("CanBePrimaryConstructorProperty")
+    public var array: ByteArray = array
```

Comment:
why public mutability?
and why at all it should be mutable?
Also, after mutating pooling will not work